ddsrouter_utils/include/ddsrouter_utils/memory/OwnerPtr.hpp

```diff
@@ -0,0 +1,320 @@
+// Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file OwnerPtr.hpp
+ *
+ * This file contains class OwnerPtr implementation.
+ */
+
+#ifndef _DDSROUTERUTILS_MEMORY_OWNERPTR_HPP_
+#define _DDSROUTERUTILS_MEMORY_OWNERPTR_HPP_
+
+#include <memory>
+#include <mutex>
+
+#include <ddsrouter_utils/library/library_dll.h>
+
+namespace eprosima {
+namespace ddsrouter {
+namespace utils {
+
+template <class T>
+class OwnerPtr;
+
+/**
+ * @brief This classes implement a new smart pointer that allows to reference data with only one ownership
+ * from different scopes and threads.
+ *
+ * INSTRUCTIONS
+ * The correct way of using it is having a \c OwnerPtr object referencing the data, and getting
+ * a lessee from it.
+ * Before using the data in the ptr, use \c lock and check the result is not null.
+ * Destroy the ptr after using it. It is advisable to destroy it soon, as the owner could want to erase it.
+ *
+ * USE CASE
+ * Whenever an object must be created and deleted from a single point, but could be used from different
+ * scopes and/or threads.
+ *
+ * WHY NEW CLASS
+ * It is similar to a shared_ptr and weak_ptr, but this assures the object is only destroyed from its
+ * owner, and only if any lessee is not using it at the moment.
+ *
+ * IMPLEMENTATION
+ * The way it works, is that every \c OwnerPtr shares a mutex with every \c LesseePtr , so while the
+ * data is being used by a lessee, the owner could not destroy it.
+ * Whenever the object is used from a lessee, a new ptr is created and mutex lock. When this ptr is deleted
+ * the mutex is unlocked (by unique_ptr deleter).
+ *
+ *
+ * EXAMPLE OF USAGE
+ * OwnerPtr<T> owner(new T());
+ * LesseePtr<T> lessee = owner.lease();
+ * {
+ *   auto locked_ptr_ = lessee.lock(); // From here till locked_ptr_ destruction, data could not be destroyed.
+ *   if (locked_ptr_) {
+ *     locked_ptr_->foo(); // Use data without worrying it could be destroyed
+ *   } else {
+ *     // Ups, the data is not available anymore
+ *   }
+ * }
+ */
+
+/**
+ * This class contains a reference (ptr) to a value of type T but do not own it.
+ * Not owning the object implies that, when using, this reference could have been invalidated.
+ * For that, method \c lock create a smart pointer referencing that object, that will not be destroyed
+ * while the smart pointer exists.
```

Comment:
```suggestion
 * In order to avoid that, method \c lock creates a smart pointer referencing that object, that will not be destroyed
 * while the smart pointer exists.
```